### ISSUE TITLE
predicted_prevalence/estimate_effect spline CIs depend on covariate scale (inflate + asymmetric on large/uncentered covariates)

### ISSUE THREAD

**Issue by nealcaren** [nealcaren/topica#882]
### Summary
For a `spline(x, df=...)` prevalence term, `predicted_prevalence(..., continuous=...)` returns confidence bands whose width depends on the covariate's **numeric scale**, not just the data. On a large, uncentered covariate (e.g. `day` ∈ [0, 3500]) the band **blows up (~10–25×) toward the high end and becomes strongly asymmetric**, even when document density is uniform over the range. Rescaling the covariate (`day/1000`, or z-scoring) — a mathematical no-op — collapses the band and restores symmetry. **Point estimates are identical either way.**

This misleads any time-spline / `~ s(day)` workflow: the CI reads as "we know almost nothing at the end of the series" when the data actually pins the estimate down tightly.

### Reproducer (self-contained, ~15s)
`topica_ci_repro.py` fits a tiny STM on a synthetic corpus with **uniform** document density over `day ∈ [0,3500]` and sweeps the same spline via `predicted_prevalence`, once on `day` and once on `day/1000`:

```
covariate   cond(XtX)   est(mid)   CI width  left / mid / right   right/left
day         1.50e+08     +0.675   0.0234 / 0.0759 / 0.2121         9.1
day_k       1.06e+07     +0.675   0.0234 / 0.0150 / 0.0213         0.9
```

Same fit, same grid, **identical point estimate (+0.675)**; only the covariate scale differs. Raw `day` inflates the mid-range band 5× and the right end 10×, and skews it (right/left = 9.1) although nothing in the (uniform) data justifies an asymmetry.

### Real-world impact
Found while fitting an article-level STM of a 1921–1930 newspaper (~33.5k docs, `~ spline(day, df=6)`, uniform ~1,500 docs / 6-month bin). Raw-`day` bands fanned out to `right/left ≈ 24` and a right-edge 95% half-width of ~0.06 — while the **empirical** per-bin SE of the same topic proportion is ~0.002–0.006 and flat across time. i.e. the reported CI at the series end was ~30× the actual sampling uncertainty. `day/1000` matched the empirical SE.

### What it's **not**
- Not data sparsity: density is uniform in the repro (and ~flat in the real corpus).
- Not the point estimate: identical under rescaling.
- Not the Rubin `between`-draw term: reproduces at `nsims=1`.
- Not a plain `pinv(XᵀX)` mean-variance conditioning loss: a model-free `sigma²·xᵀpinv(XᵀX)x` (and a QR-stable equivalent) are both scale-invariant at cond ~1e8 in float64. So the scale sensitivity is entering somewhere in the **spline continuous-sweep variance path** specifically (`_build_reference_rows` grid-basis eval → `predicted_prevalence` `within_scale * XtX_inv` applied to `X_new`), not in the generic OLS variance.

### Likely root cause
`spline()` builds its basis on the **raw, uncentered** covariate: `cols = [x]` (the raw linear term) plus cubic terms `~ (x−t)³ / (t_max−t_min)²`. With `x` ~3500 this makes `XᵀX` ill-conditioned (cond ≈ 1.5e8 vs ~1e7 rescaled), and the grid/`X_new` variance evaluation amplifies it at the high-magnitude end.

### Suggested fixes (any one)
1. **Standardize inside `spline()`** — center/scale `x` (store shift+scale, reuse for the prediction grid & captured knots) so the basis is well-conditioned regardless of the covariate's units. Cleanest; invisible to callers.
2. Compute the OLS covariance in `_pooled_coefficients` / the sweep from a **QR (or SVD) of `X`** rather than `pinv(XᵀX)`, and apply it to `X_new` in that stable basis.
3. At minimum, **document** that spline covariates should be scaled to O(1) and warn when `cond(XᵀX)` is large.

### Workaround
Scale the covariate to O(1) before `spline()` (e.g. `day/1000` or z-score) and map the grid back for plotting. Point estimates unaffected; CIs then match the empirical SE.

Env: topica 0.59.0 (dev), Python 3.13, macOS.

**Comment by nealcaren** [comment]
Full reproducer referenced above (`topica_ci_repro.py`):

```python
"""Self-contained repro: predicted_prevalence spline CIs depend on the covariate's
SCALE -- they blow up (~10x) and turn asymmetric on a large/uncentered covariate,
even though rescaling is a mathematical no-op and the point estimates are identical.

Same fitted model, same data, same grid; only `day` (0..3500) vs `day/1000` differs.
Run: python topica_ci_repro.py
"""
import itertools, string
import numpy as np, pandas as pd, topica

rng = np.random.default_rng(0)
D, V, K = 1200, 60, 4
vocab = ["".join(t) for t in itertools.islice(itertools.product(string.ascii_lowercase, repeat=3), V)]

# uniform document density over a large, uncentered covariate ("days since start")
day = np.sort(rng.uniform(0, 3500, D))
w = 0.5 + 0.4 * np.sin(day / 700)                 # topic-0 prevalence, smooth in time
theta = np.zeros((D, K)); theta[:, 0] = w; theta[:, 1:] = (1 - w)[:, None] / (K - 1)
topic_word = rng.dirichlet(np.ones(V) * 0.3, size=K)
docs = []
for d in range(D):
    zs = rng.choice(K, size=rng.integers(40, 80), p=theta[d])
    docs.append(" ".join(vocab[rng.choice(V, p=topic_word[z])] for z in zs))

df = pd.DataFrame({"text": docs, "day": day, "day_k": day / 1000.0})
corpus = topica.from_dataframe(df, text_col="text", metadata_cols=["day", "day_k"],
                               min_length=1, min_doc_freq=1)
corpus, meta = topica.prep_documents(corpus, lower_thresh=1)
meta = pd.DataFrame(meta)

b, names = topica.spline(meta["day"].to_numpy(), df=6)
model = topica.STM(K, seed=0)
model.fit(corpus, prevalence=b, prevalence_names=names, iters=150)

print(f"{'covariate':8}  {'cond(XtX)':>10}  {'est(mid)':>9}   CI width  left / mid / right   right/left")
for col in ("day", "day_k"):
    e = topica.predicted_prevalence(model, formula=f"~ spline({col}, df=6)", data=meta,
                                    continuous=col, npoints=40, ci=0.95, seed=0, topics=[0])[0]
    est = np.array(e.estimate); wdt = np.array(e.ci_high) - np.array(e.ci_low)
    B, _ = topica.spline(meta[col].to_numpy(), df=6)
    X = np.hstack([np.ones((len(meta), 1)), B]); cond = np.linalg.cond(X.T @ X)
    print(f"{col:8}  {cond:10.2e}  {est[20]:+9.3f}   "
          f"{wdt[0]:.4f} / {wdt[20]:.4f} / {wdt[-1]:.4f}       {wdt[-1]/wdt[0]:5.1f}")
print("\nPoint estimates are identical; only the CI differs, purely from covariate scale.")
```

**Comment by nealcaren** [comment]
### How R `stm` avoids this (points squarely at the QR fix)

I checked what R's `stm` does at this exact spot, since `predicted_prevalence`'s docstring says it mirrors `plot.estimateEffect`. R `stm` **does not scale the covariate** — it uses raw `s(day)` / `bs()` bases via `model.matrix` — yet it never exhibits this blow-up. The reason is purely numerical: it never forms `XᵀX`.

`stm::estimateEffect`:
```r
qx <- qr(xmat)                       # QR of X  (cond ~1e4 for raw day)
lm.mod <- qr.lm(thetasims[,k], qx)   # solve.qr(qx, y)
summary.qr.lm(lm.mod)
```
`stm:::summary.qr.lm`:
```r
Qr <- object$qr
R <- chol2inv(Qr$qr[p1, p1, drop = FALSE])   # (XᵀX)^-1 from the QR's R factor
list(est = ..., vcov = sigma^2 * R)          # coefficient covariance
```
`plot.estimateEffect` then simulates coefficients from this vcov via `rmvnorm` and pushes them through the design grid.

The key difference: R gets `(XᵀX)^-1` as `chol2inv(R)` from the **QR of `X`**, whereas topica does `np.linalg.pinv(X.T @ X)`. **Forming `XᵀX` squares the condition number** (here 1e4 → 1e8), so `pinv` loses ~8 digits precisely in the high-magnitude (right-edge) directions of an unscaled spline basis — which is the asymmetric inflation in this report. QR on `X` only ever sees cond ~1e4.

So **fix option 2 (QR/`chol2inv` instead of `pinv(XᵀX)`) is what R already does**, and it makes topica robust to covariate scale without requiring callers to standardize — matching `stm`'s behavior. Standardizing inside `spline()` (option 1) is still a nice belt-and-suspenders, but the QR change is the one that closes the gap with R.

**Comment by nealcaren** [comment]
## Fixed in #883 (shipped in 0.60.0) — root cause was the Cholesky jitter, not basis conditioning

Fixed, with a small correction to the diagnosis. This issue localized it correctly to the
sampling/Sigma path and correctly ruled out the quadratic form (`xᵀpinv(XtX)x` *is*
scale-invariant at cond ~1e8 — confirmed), but pinned the cause on the ill-conditioned
`spline()` basis. The actual culprit is one line:

`Sigma_sym = 0.5*(Σ+Σ.T) + 1e-10*np.eye(p)` (the fixed absolute Cholesky jitter).

Because the band is drawn as `eta = beta_draws @ X_new.T`, that jitter contributes exactly
`1e-10 · ‖X_new[g]‖²` to each grid point's variance — which is **not** scale-invariant. For
raw `day`, `‖X_new‖² ≈ 2.9e7` and grows with the covariate, so the jitter term (~2.9e-3)
buries the true variance (~6e-6): ~471× inflation, right-heavy. For `day/1000`,
`‖X_new‖² ≈ 30` → jitter term 3e-9, negligible. That's the whole bug — the ill-conditioned
`XᵀX` is a red herring (the quadratic form handles cond 1e8 fine).

**Fix (#883):** form the coefficient-draw square root from the eigendecomposition (clipping
tiny negative eigenvalues) with no additive jitter — this was already the code's `except`
fallback; it needs no regularizer and is scale-invariant. No basis standardization needed.

**Verification:** isolated, raw-`day` right/left goes 486.98 → 1.03 (= the `day/1000` value);
on a real STM spline fit the raw and rescaled bands now match (allclose) with a symmetric
max/min ~2.5 (the legitimate spline end-variance) instead of the 10–25× right blowup. Point
estimates unchanged. Regression test `test_spline_ci_is_scale_invariant` guards it.

Scope: only `predicted_prevalence`'s coefficient-draw covariance. The per-document η samplers
keep their jitter — η is O(1), so `1e-10·‖η‖²` is harmless there; and `s()`/`bs()` bases are
bounded O(1), so only the natural-cubic `spline()` term ever triggered this.
